src/algorithms/calorimetry/CalorimeterClusterRecoCoG.cc

```diff
@@ -338,4 +295,169 @@ std::optional<edm4eic::MutableCluster> CalorimeterClusterRecoCoG::reconstruct(co
+          mapMCParToSimIndices.insert(
+            {idPrim, std::make_pair(iAssocSimHit, iContrib)}
+          );
+          mapMCParToContrib.insert(
+            {idPrim, contrib.getEnergy()}
+          );
+        } else {
+          mapMCParToContrib[idPrim] += contrib.getEnergy();
+        }
+        trace("Identified primary: id = {}, pid = {}, total energy = {}, contributed = {}",
+          primary.getObjectID().index,
+          primary.getPDG(),
+          primary.getEnergy(),
+          mapMCParToContrib[idPrim]
+        );
+
+      }  // end contribution loop
+    }  // end associated sim hit loop
+  }  // end hit loop
+  debug("Found {} primaries contributing a total of {} GeV", mapMCParToContrib.size(), eSimHitSum);
+
+  // --------------------------------------------------------------------------
+  // 3. create association for each contributing primary
+  // --------------------------------------------------------------------------
+  for (const auto& parAndSimIndices : mapMCParToSimIndices) {
+
+    // grab indices, calculate weight
+    const uint32_t iAssocSimHit = parAndSimIndices.second.first;
+    const uint32_t iContrib = parAndSimIndices.second.second;
+    const double weight = mapMCParToContrib[parAndSimIndices.first] / eSimHitSum;
+
+    // get relevant MCParticle
+    edm4hep::MCParticle primary;
+#if EDM4EIC_VERSION_MAJOR >= 7
+    get_primary((*mchitassociations)[iAssocSimHit].getSimHit().getContributions(iContrib), primary);
+#else
+    get_primary((*mchits)[iAssocSimHit].getContributions(iContrib), primary);
+#endif
+
+    // set association
+    auto assoc = assocs->create();
+    assoc.setRecID(cl.getObjectID().index); // if not using collection, this is always set to -1
+    assoc.setSimID(primary.getObjectID().index);
+    assoc.setWeight(weight);
+    assoc.setRec(cl);
+    assoc.setSim(primary);
+    debug("Associated cluster #{} to MC Particle #{} (pid = {}, status = {}, energy = {}) with weight ({})",
+      cl.getObjectID().index,
+      primary.getObjectID().index,
+      primary.getPDG(),
+      primary.getGeneratorStatus(),
+      primary.getEnergy(),
+      weight
+    );
+
+  }  // end contributor loop
+  return;
+
+}  // end 'associate(edm4eic::Cluster&, edm4eic::MCRecoCalorimeterHitAssocationCollection OR edm4hep::SimCalorimeterHitCollection*, edm4eic::MCRecoClusterParticleAssociationCollection*)'
+
+//------------------------------------------------------------------------
+void CalorimeterClusterRecoCoG::get_primary(
+  const edm4hep::CaloHitContribution& contrib,
+  edm4hep::MCParticle& primary
+) const {
+
+  // get contributing particle
+  const auto contributor = contrib.getParticle();
+
+  // walk back through parents to find primary
+  //   - TODO finalize primary selection. This
+  //     can be improved!!
+  primary = contributor;
+  while (primary.parents_size() > 0) {
+    primary = primary.getParents(0);
+    if (primary.getGeneratorStatus() == 1) break;
+  }
+  return;
```

Comment:
```suggestion
  return primary;
```
Return the proxy object, don't need to update it by reference.